### PR TITLE
Recycle all logical replication messages

### PR DESCRIPTION
**Pull request by roji**
Based on top of #5 - filter out first commit when reviewing.

Also redo TupleData as a lightweight discriminated union

Note: We now expose ReadOnlyMemory<T> instead of T[], both for immutability and recycling (slice over a larger backing array). I had the unpleasant surprise that Memory and ROM don't have simple indexers, so to get an element you need to get a Span from them :/ Not sure why it's like that. ArraySegment is also a struct-based array slice, but does not provide immutability. We can consider if there's a better option.

Note: this leaves out physical and test_decoding messages for since it's late :sle

### CODE REVIEW

**Review by Brar — APPROVED** [review]
LGTM modulo naming choice for the slot (see #5)

**Comment by roji** [comment]
Thanks. Will hold off merging this until #5 is merged.

**Comment by roji** [comment]
Merged via a3e965acc9844a5588397a1afdbcd4fc133ecaa5

**Pull request closed without merging**
